crates/sncast/src/starknet_commands/verify/voyager.rs

```diff
@@ -0,0 +1,355 @@
+    let package_root = &package_metadata.root;
+
+    if let Some(lic) = package_metadata
+        .manifest_metadata
+        .license_file
+        .as_ref()
+        .map(Utf8Path::new)
+        .map(Utf8Path::to_path_buf)
+    {
+        sources.push(package_root.join(lic));
+    }
+
+    if let Some(readme) = package_metadata
+        .manifest_metadata
+        .readme
+        .as_deref()
+        .map(Utf8Path::new)
+        .map(Utf8Path::to_path_buf)
+    {
+        sources.push(package_root.join(readme));
+    }
+
+    Ok(sources)
+}
+
+fn biggest_common_prefix<P: AsRef<Utf8Path> + Clone>(
+    paths: &[Utf8PathBuf],
+    first_guess: P,
+) -> Utf8PathBuf {
+    let ancestors = Utf8Path::ancestors(first_guess.as_ref());
+    let mut biggest_prefix: &Utf8Path = first_guess.as_ref();
+    for prefix in ancestors {
+        if paths.iter().all(|src| src.starts_with(prefix)) {
+            biggest_prefix = prefix;
+            break;
+        }
+    }
+    biggest_prefix.to_path_buf()
+}
+
+impl Voyager {
+    fn gather_files(&self) -> Result<(Utf8PathBuf, HashMap<String, Utf8PathBuf>)> {
+        let mut packages = vec![];
+        gather_packages(&self.metadata, &mut packages)?;
+
+        let mut sources: Vec<Utf8PathBuf> = vec![];
+        for package in &packages {
+            let mut package_sources = package_sources(package)?;
+            sources.append(&mut package_sources);
+        }
+
+        let prefix = biggest_common_prefix(&sources, &self.metadata.workspace.root);
+        let manifest_path = &self.metadata.workspace.manifest_path;
+        let manifest = manifest_path
+            .strip_prefix(&prefix)
+            .map_err(|_| anyhow!("Couldn't strip {prefix} from {manifest_path}"))?;
+
+        let mut files: HashMap<String, Utf8PathBuf> = sources
+            .iter()
+            .map(|p| -> Result<(String, Utf8PathBuf)> {
+                let name = p
+                    .strip_prefix(&prefix)
+                    .map_err(|_| anyhow!("Couldn't strip {prefix} from {p}"))?;
+                Ok((name.to_string(), p.clone()))
+            })
+            .try_collect()?;
+        files.insert(
+            manifest.to_string(),
+            self.metadata.workspace.manifest_path.clone(),
+        );
+
+        Ok((prefix, files))
+    }
+}
+
+#[async_trait::async_trait]
+impl VerificationInterface for Voyager {
+    fn new(network: Network, workspace_dir: Utf8PathBuf) -> Result<Self> {
+        let manifest_path = scarb_utils::get_scarb_manifest_for(workspace_dir.as_ref())?;
+        let metadata = scarb_utils::get_scarb_metadata_with_deps(&manifest_path)?;
+        let url_str = match env::var("STARKNET_RPC_URL") {
+            Ok(addr) => addr,
+            Err(_) => match network {
+                Network::Mainnet => "https://free-rpc.nethermind.io/mainnet-juno/v0_7".to_string(),
+                Network::Sepolia => "https://free-rpc.nethermind.io/sepolia-juno/v0_7".to_string(),
+            },
+        };
```

Comment:
1. We don't have any mechanism which allows to set rpc url through env.
2. Have a look at `Network` impl, it provides methods which return provider url 😄 .
3. This won't apply if you follow step 2., but we should use rpc 0.8

Reply:
> Have a look at Network impl, it provides methods which return provider url 😄 .

Thanks, that seemed like something I was missing. You're referring to [this one](https://github.com/foundry-rs/starknet-foundry/blob/034c433c1ee2d5e6897179b0cb3378a32e0b3f9f/crates/sncast/src/helpers/rpc.rs#L84)? All those functions are private.

Looking at that module perhaps I should add `RpcArgs` to verify?

Reply:
> Looking at that module perhaps I should add RpcArgs to verify?

Yes, that should solve the issue.

Reply:
Unfortunately I couldn't reuse much of existing machinery, due to how `--network` interacts with `CastConfig`. Please see how it looks now. I think the test failure is unrelated and needs a restart.